src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,180 @@
+
+    private static class Node<T> {
+        private T item;
+        private Node<T> next;
+        private Node<T> prev;
+
+        Node(Node<T> prev, T element, Node<T> next) {
+            this.item = element;
+            this.next = next;
+            this.prev = prev;
+        }
+    }
+
     @Override
     public void add(T value) {
+        if (size == 0) {
+            first = new Node<>(null, value, null);
+            last = first;
+        } else {
+            last.next = new Node<>(last, value, null);
+            last.next.prev = last;
+            last = last.next;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index < 0 || index > size) {
+            checkIndexValidity(index);
+        }
+        if (index == 0 && size > 0) {
+            addToTheBeginning(value);
+        }
+        if (index == size) {
+            addToTheEnd(value);
+        } else if (index > 0) {
+            addToTheMiddle(value, index);
+        }
+    }
+
+    private void addToTheMiddle(T value, int index) {
+        int i = 0;
+        for (MyLinkedList.Node<T> x = first; x != null; x = x.next) {
+            if (i == index) {
+                Node<T> newNode = new Node<>(x.prev, value, x);
+                x.prev.next = newNode;
+                x.prev = newNode;
+                break;
+            }
+            i++;
+        }
+        size++;
+    }
+
+    private void addToTheEnd(T value) {
+        add(value);
+    }
+
+    private void addToTheBeginning(T value) {
+        Node<T> newNode = new Node<>(null, value, first);
+        first.prev = newNode;
+        first = newNode;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T item: list) {
+            add(item);
+        }
     }
 
     @Override
     public T get(int index) {
+        checkIndexValidity(index);
+        int i = 0;
+        for (MyLinkedList.Node<T> x = first; x != null; x = x.next) {
+            if (i == index) {
+                return x.item;
+            }
+            i++;
+        }
         return null;
     }
 
+    private void checkIndexValidity(int index) {
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException();
+        }
+    }
+
     @Override
     public T set(T value, int index) {
+        checkIndexValidity(index);
+        int i = 0;
+        for (MyLinkedList.Node<T> x = first; x != null; x = x.next) {
+            if (i == index) {
+                T oldValue = x.item;
+                x.item = value;
+                return oldValue;
+            }
+            i++;
+        }
         return null;
     }
 
     @Override
     public T remove(int index) {
+        checkIndexValidity(index);
+        int i = 0;
+        for (MyLinkedList.Node<T> x = first; x != null; x = x.next) {
+            if (i == index) {
+                T removedItem = x.item;
+                remove(x.item);
```

Comment:
You are calling the `remove(T object)` method inside the `remove(int index)` method which makes the time complexity of this method O(n^2). Instead, you should unlink the node directly here.